lib/parser.dart

```diff
@@ -1033,8 +1042,85 @@ class _Parser {
     _eat(TokenKind.LBRACE);
     var groupRuleBody = processGroupRuleBody();
     _eat(TokenKind.RBRACE);
-    return new DocumentDirective(
-        functions, groupRuleBody, _makeSpan(start));
+    return new DocumentDirective(functions, groupRuleBody, _makeSpan(start));
+  }
+
+  SupportsDirective processSupportsDirective() {
+    var start = _peekToken.span;
+    _next(); // '@supports'
+    var condition = processSupportsCondition();
+    _eat(TokenKind.LBRACE);
+    var groupRuleBody = processGroupRuleBody();
+    _eat(TokenKind.RBRACE);
+    return new SupportsDirective(condition, groupRuleBody, _makeSpan(start));
+  }
+
+  SupportsCondition processSupportsCondition() {
+    var kind = _peek();
+    if (kind == TokenKind.IDENTIFIER) {
+      return processSupportsNegation();
+    }
+
+    var start = _peekToken.span;
+    var conditions = <SupportsConditionInParens>[];
+    var clauseType = ClauseType.none;
+
+    while (true) {
```

Comment:
Perhaps I'm misunderstanding your point, but the intent here is to collect all terms in a conjunction _or_ disjunction. The grammar requires parentheses around any mixed clauses, so there are no precedence issues to worry about.

Could you elaborate on your suggestion? Are you suggesting treating the operators in their typical binary sense recursively? So that `A and B and C` would be parsed as:
```
SupportsConjunction: [
  A,
  SupportsConjunction: [
    B,
    C,
  ]
]
```

Such an approach wouldn't work because the terms in a conjunction must be a `SupportsConditionInParens`. Or are you just suggesting that I split all the `and`/`or` logic out into two separate functions for processing conjunctions and disjunctions? I think you'd still need a loop to collect all the sequential terms in the clause.